### PR TITLE
add deadline and countdown to application section

### PR DESCRIPTION
**Pull request by nmoskaleva**

### CODE REVIEW

File: components/apply-section/deadline/deadline.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    marginBottom: 0,
+    padding: 0
+  }
+})
+
+const classNumber = 15
+const deadlineDate = '02 / 20 / 2020'
+const classStartDate = '03 / 01 / 2020'
+
+const formattedDeadlineDate = new Date(deadlineDate).toLocaleString('en', {
```
Comment (benna100) [inline components/apply-section/deadline/deadline.js]: What do you think about applicationEndDate?

File: components/apply-section/deadline/deadline.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  day: 'numeric',
+  month: 'long',
+  year: 'numeric'
+})
+
+const formattedNewClassDate = new Date(classStartDate).toLocaleString('en', {
```
Comment (benna100) [inline components/apply-section/deadline/deadline.js]: And applicationStartDate?

File: components/apply-section/deadline/deadline.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  year: 'numeric'
+})
+
+export default function deadline() {
+  const classes = useStyles()
+
+  return (
+    <React.Fragment>
+      <Typography className={classes.deadline}>
+        <span className={classes.deadlineClass}>Class {classNumber}</span>{' '}
```
Comment (benna100) [inline components/apply-section/deadline/deadline.js]: Where does the deadlineClass come from? And why the {' '}?

File: components/apply-section/deadline/deadline.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        starts on {formattedNewClassDate} | Application Deadline:{' '}
+        {formattedDeadlineDate}
+      </Typography>
+      <Container>
+        <Timer date={formattedDeadlineDate} />
```
Comment (benna100) [inline components/apply-section/deadline/deadline.js]: Maybe dateToCountDownFrom? Not great, but maybe you have a better suggestion. countDownDate maybe

File: components/apply-section/deadline/timer.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    }
+  }
+}))
+
+export default function timer(props) {
+  const classes = useStyles()
+  return (
+    <div className={classes.root}>
+      <List className={classes.list}>
+        {countdown(props.date).map(item => (
```
Comment (benna100) [inline components/apply-section/deadline/timer.js]: Too much is going on here, so its quite hard to follow the code here. 
Try break it up, help me as a developer understand the flow, fx:
const dateToCountDownFrom= props.date;
const countDownComponents = countdown(dateToCountDownFrom);
countDownComponents.map(...)

Never call a variable item or data, etc. Its too vague. countDownComponent would maybe be better?

Also maybe deconstruct item? [primary, secondary]

File: components/apply-section/deadline/timer.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+          <ListItem className={classes.listItem} key={item}>
+            <ListItemText
+              classes={{
+                primary: classes.primaryText,
+                secondary: classes.secondaryText
+              }}
+              primary={item[0].split(' ', 1)}
```
Comment (benna100) [inline components/apply-section/deadline/timer.js]: We are talking countdown date and here i have something called primary and secondary i have no idea what that is. Also i would expect the countdown(dateToCountDownFrom) to return something i dont have to split up.

File: components/apply-section/deadline/countdown.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  const [timeLeft, setTimeLeft] = useState(calculateTimeLeft())
+
+  useEffect(() => {
+    const intervalId = setInterval(() => {
+      setTimeLeft(calculateTimeLeft())
+    }, 1000)
+
+    return () => clearInterval(intervalId)
```
Comment (benna100) [inline components/apply-section/deadline/countdown.js]: This is tricky to read. Im not following the flow or what you are trying to achieve. Its a function that return an array of the states key on an item which is string with a space. 

Again help me as a developer understand the flow!
Fx TimeLeft for me sounds like a number, but its an object with lots of keys, why?
Why are you iterating over the keys? And when you are, make sure to give a descriptive parameter name, item is no go. Maybe dateComponent??? And then why are you pushing an array with one element that is split up with a space? What about saying:

countdownComponents.push({item: timeLeft[item]})
Then you can use it like countdownComponents[0].days.

File: components/apply-section/deadline/countdown.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import React, { useState, useEffect } from 'react'
+
+export default function countdown(date) {
+  const calculateTimeLeft = () => {
+    const diff = +new Date(date) - +new Date()
```
[the remark this answers is not in the document]
Comment (nmoskaleva) [reply components/apply-section/deadline/countdown.js]: > I was especially confused by countdown.js, why not just return the rendered date from that function instead of the date components?

I'll explain what happened. timeLeft is an object (and then array) because of the styling basically. I know it's a bad practice, but I tried to make it look as columns of number + time unit, and didn't see how to do it with a date string. 

So Material-UI has a ListItemText component which I thought would be appropriate. It has a primary text (larger) and a secondary text (smaller). So one list item is in this case [primary text + secondary text]. For example [15 + seconds]. 
I needed to map those two together in timer.js, because they are in total 4 list items, not one date/ string.  
But also I thought timeLeft doesn't look too bad as an object. I renamed it to timeLeftInUnits. Is it better maybe?

Or we can make it a string and render in a different way.

File: components/apply-section/deadline/deadline.js (hunk 1 above)
Comment (nmoskaleva) [reply components/apply-section/deadline/deadline.js]: yes!

File: components/apply-section/deadline/deadline.js (hunk 2 above)
Comment (nmoskaleva) [reply components/apply-section/deadline/deadline.js]: newClassStartDate i think, it's an approx. date when new class will start

File: components/apply-section/deadline/deadline.js (hunk 4 above)
Comment (nmoskaleva) [reply components/apply-section/deadline/deadline.js]: applicationEndDate

File: components/apply-section/deadline/timer.js (hunk 5 above)
Comment (nmoskaleva) [reply components/apply-section/deadline/timer.js]: changed

File: components/apply-section/deadline/timer.js (hunk 6 above)
Comment (nmoskaleva) [reply components/apply-section/deadline/timer.js]: primary and secondary are names of classes in material-ui that allow overriding the style of the component. But making strings with white space and then splitting them – unexplainable. Sorry. Something went wrong at that moment.

File: components/apply-section/deadline/countdown.js (hunk 7 above)
Comment (nmoskaleva) [reply components/apply-section/deadline/countdown.js]: thanks a lot for the review, Benjamin!

I'll explain what happened. timeLeft is an object (and then array) because of the styling basically. I know it's a bad practice, but I tried to make it look as columns of number + time unit, and didn't see how to do it with a date string. 

So Material-UI has a ListItemText component which I thought would be appropriate. It has a primary text (larger) and a secondary text (smaller). So one list item is in this case [primary text + secondary text]. For example [15 + seconds]. 
I needed to map those two together in timer.js, because they are in total 4 list items, not one date/ string.  
But also I thought timeLeft doesn't look too bad as an object. I renamed it to timeLeftInUnits. Is it better maybe?

Or we can make it a string and render in a different way.

File: components/apply-section/deadline/deadline.js (hunk 3 above)
Comment (nmoskaleva) [reply components/apply-section/deadline/deadline.js]: {' '} deleted (was Prettier for white space )

File: components/apply-section/deadline/timer.js (hunk 6 above)
Comment (benna100) [reply components/apply-section/deadline/timer.js]: No worries! But cool, good to know!

File: components/apply-section/deadline/countdown.js (hunk 7 above)
Comment (benna100) [reply components/apply-section/deadline/countdown.js]: These kind of considerations is why am always a bit reluctant to use material-ui. In my opinion it does not provide that much value, but it sure seems to constraint us quite a bit. 

Off course these things are not black and white, but lets just keep going with it.

File: components/apply-section/deadline/countdown.js [hunk 9]
```diff
@@ -1,42 +1,25 @@
...
 
-  const [timeLeft, setTimeLeft] = useState(calculateTimeLeft())
+export default function countdown(date) {
+  const [timeLeftInUnits, setTimeLeftInUnits] = useState(getTimeToDeadline())
 
   useEffect(() => {
-    const intervalId = setInterval(() => {
-      setTimeLeft(calculateTimeLeft())
+    const oneSecondInterval = setInterval(() => {
+      setTimeLeftInUnits(getTimeToDeadline(date))
```
Comment (benna100) [inline components/apply-section/deadline/countdown.js]: Ahh yes, now this is nice to read!

File: components/apply-section/deadline/helpers.js [hunk 10]
```diff
@@ -0,0 +1,24 @@
+// calculate time difference between today and the deadline
+const getTimeToDeadline = date => {
```
Comment (benna100) [inline components/apply-section/deadline/helpers.js]: Simple function that only does one specific thing. Really really nice!!

File: components/apply-section/deadline/timer.js [hunk 11]
```diff
@@ -52,25 +49,29 @@ const useStyles = makeStyles(theme => ({
...
+export default function Timer(props) {
   const classes = useStyles()
+  const dateToCountDownFrom = props.date
+  const countDownComponents = countdown(dateToCountDownFrom)
   return (
     <div className={classes.root}>
       <List className={classes.list}>
-        {countdown(props.date).map(item => (
-          <ListItem className={classes.listItem} key={item}>
+        {countDownComponents.map(timeUnit => (
```
Comment (benna100) [inline components/apply-section/deadline/timer.js]: Yes, now i can follow this a lot easier
